File: hilbert-curve-sort.js
export async function h2CurveSort(vec2s) {
	let [minX, minY] = ([maxX, maxY] = vec2s[0]);
	for (let i = 1; i < vec2s.length; i++) {
		const [x, y] = vec2s[i];
		if (x < minX) minX = x;
		else if (x > maxX) maxX = x;
		if (y < minY) minY = y;
		else if (y > maxY) maxY = y;
	}
	const sideX = maxX - minX;
	const sideY = maxY - minY;
	const maxSide = Math.max(sideX, sideY);

	// Centering and scaling data to fit hypercube
	let scaleX = maxSide / sideX;
	let scaleY = maxSide / sideY;
	if (scaleX === 0 || !Number.isFinite(scaleX)) scaleX = 1;
	if (scaleY === 0 || !Number.isFinite(scaleY)) scaleY = 1;
	const normVec2s = vec2s.map(([x, y]) => [scaleX * (x - minX), scaleY * (y - minY)]);

	// De-scaling and de-centering the results from actual sort
	return (await _h2CurveSort(normVec2s, maxSide)).map(([x, y]) => [x / scaleX + minX, y / scaleY + minY]);
}

export async function h3CurveSort(vec3s) {
	let [minX, minY, minZ] = ([maxX, maxY, maxZ] = vec3s[0]);
	for (let i = 1; i < vec3s.length; i++) {
		const [x, y, z] = vec3s[i];
		if (x < minX) minX = x;
		else if (x > maxX) maxX = x;
		if (y < minY) minY = y;
		else if (y > maxY) maxY = y;
		if (z < minZ) minZ = z;
		else if (z > maxZ) maxZ = z;
	}
	const sideX = maxX - minX;
	const sideY = maxY - minY;
	const sideZ = maxZ - minZ;
	const maxSide = Math.max(sideX, sideY, sideZ);

	// Centering and scaling data to fit hypercube
	let scaleX = maxSide / sideX;
	let scaleY = maxSide / sideY;
	let scaleZ = maxSide / sideZ;
	if (scaleX === 0 || !Number.isFinite(scaleX)) scaleX = 1;
	if (scaleY === 0 || !Number.isFinite(scaleY)) scaleY = 1;
	if (scaleZ === 0 || !Number.isFinite(scaleZ)) scaleZ = 1;
	const normVec3s = vec3s.map(([x, y, z]) => [scaleX * (x - minX), scaleY * (y - minY), scaleZ * (z - minZ)]);

	// De-scaling and de-centering the results from actual sort
	return (await _h3CurveSort(normVec3s, maxSide)).map(([x, y, z]) => [
		x / scaleX + minX,
		y / scaleY + minY,
		z / scaleZ + minZ,
	]);
}

// Gray Code defining orthants' order
const grayCode = (n) => [...Array(2 ** n).keys()].map((bit) => bit ^ (bit >> 1));
const GRAY_2 = grayCode(2);
const GRAY_3 = grayCode(3);

async function _h2CurveSort(vec2s, side) {
	// Base
	if (vec2s.length < 2 || new Set(vec2s.map(String)).size === 1) return vec2s;

	// Recursion
	const mid = side / 2;
	// Transformation of quadrants to (from) U_2(1)
	const maps = {
		0b00: ([x, y]) => [y, x],
		0b01: ([x, y]) => [x, y - mid],
		0b11: ([x, y]) => [x - mid, y - mid],
		0b10: ([x, y]) => [mid - y, side - x],
	};
	const inverseMaps = {
		0b00: ([x, y]) => [y, x],
		0b01: ([x, y]) => [x, y + mid],
		0b11: ([x, y]) => [x + mid, y + mid],
		0b10: ([x, y]) => [side - y, mid - x],
	};

	const quads = [...Array(4)].map(() => []);
	for (const vec2 of vec2s) {
		const bitX = vec2[0] > mid;
		const bitY = vec2[1] > mid;
		// Quadrant as G_2 bit
		const quad = (bitX << 1) + bitY;
		quads[quad].push(vec2);
	}

	// Transform quadrants to U_2(1)
	const sorted = await Promise.all(
		quads.map(async (quadVec2s, quad) => await _h2CurveSort(quadVec2s.map(maps[quad]), mid)),
	);
	// Order quadrants w.r.t G_2 and de-transform
	return GRAY_2.flatMap((quad) => sorted[quad].map(inverseMaps[quad]));
}

async function _h3CurveSort(vec3s, side) {
	// Base
	if (vec3s.length < 2 || new Set(vec3s.map(String)).size === 1) return vec3s;

	// Recursion
	const mid = side / 2;
	// Transformation of octants to (from) U_3(1)
	const maps = {
		0b000: ([x, y, z]) => [z, x, y],
		0b001: ([x, y, z]) => [y, z - mid, x],
		0b011: ([x, y, z]) => [y - mid, z - mid, x],
		0b010: ([x, y, z]) => [x, side - y, mid - z],
		0b110: ([x, y, z]) => [x - mid, side - y, mid - z],
		0b111: ([x, y, z]) => [side - y, z - mid, side - x],
		0b101: ([x, y, z]) => [mid - y, z - mid, side - x],
		0b100: ([x, y, z]) => [mid - z, side - x, y],
	};
	const inverseMaps = {
		0b000: ([x, y, z]) => [y, z, x],
		0b001: ([x, y, z]) => [z, x, y + mid],
		0b011: ([x, y, z]) => [z, x + mid, y + mid],
		0b010: ([x, y, z]) => [x, side - y, mid - z],
		0b110: ([x, y, z]) => [x + mid, side - y, mid - z],
		0b111: ([x, y, z]) => [side - z, side - x, y + mid],
		0b101: ([x, y, z]) => [side - z, mid - x, y + mid],
		0b100: ([x, y, z]) => [side - y, z, mid - x],
	};

	const octs = [...Array(8)].map(() => []);
	for (const vec3 of vec3s) {
		const bitX = vec3[0] > mid;
		const bitY = vec3[1] > mid;
		const bitZ = vec3[2] > mid;
		// Octant as G_3 bit
		const oct = (bitX << 2) + (bitY << 1) + bitZ;
		octs[oct].push(vec3);
	}

	// Transform octants to U_3(1)
	const sorted = await Promise.all(
		octs.map(async (octVec3s, oct) => await _h3CurveSort(octVec3s.map(maps[oct]), mid)),
	);
	// Order octants w.r.t G_3 and de-transform
	return GRAY_3.flatMap((oct) => sorted[oct].map(inverseMaps[oct]));
}

File: src/cpp/hilbert_curve_sort.cc
// hilbert_curve_sort.cc
// refer to ..\ts\hilbert-curve-sort.ts for documentation

#include "hilbert_curve_sort.h"

#include <algorithm>
#include <cstdint>
#include <functional>
#include <iterator>

void run_sort_2d(std::vector<std::array<double, 2>> &vec2s, double side);
auto is_base_2d(const std::vector<std::array<double, 2>> &vec2s) -> bool;
void run_sort_3d(std::vector<std::array<double, 3>> &vec3s, double side);
auto is_base_3d(const std::vector<std::array<double, 3>> &vec3s) -> bool;

constexpr std::array<uint8_t, 4> gray2 = {0b00, 0b01, 0b11, 0b10};
constexpr std::array<uint8_t, 8> gray3 = {0b000, 0b001, 0b011, 0b010,
                                          0b110, 0b111, 0b101, 0b100};

void sort_2d(std::vector<std::array<double, 2>> &vec2s) {
  const auto [min_x_it, max_x_it] = std::ranges::minmax_element(
      vec2s, {}, [](const auto &vec) { return vec[0]; });
  const double min_x = (*min_x_it)[0];
  const double max_x = (*max_x_it)[0];

  const auto [min_y_it, max_y_it] = std::ranges::minmax_element(
      vec2s, {}, [](const auto &vec) { return vec[1]; });
  const double min_y = (*min_y_it)[1];
  const double max_y = (*max_y_it)[1];

  const double side_x = max_x - min_x;
  const double side_y = max_y - min_y;
  const double bound = std::max(side_x, side_y);

  const double scale_x = (side_x == 0 || bound == 0) ? 1 : bound / side_x;
  const double scale_y = (side_y == 0 || bound == 0) ? 1 : bound / side_y;

  for (auto &vec2 : vec2s) {
    vec2[0] = scale_x * (vec2[0] - min_x);
    vec2[1] = scale_y * (vec2[1] - min_y);
  }

  run_sort_2d(vec2s, bound);

  for (auto &vec2 : vec2s) {
    vec2[0] = vec2[0] / scale_x + min_x;
    vec2[1] = vec2[1] / scale_y + min_y;
  }
};

void run_sort_2d(std::vector<std::array<double, 2>> &vec2s, const double side) {
  if (is_base_2d(vec2s)) {
    return;
  }

  const double mid = side / 2;

  const std::array<std::function<void(std::array<double, 2> &)>, 4> maps{
      [](auto &vec2) { vec2 = {vec2[1], vec2[0]}; },
      [mid](auto &vec2) { vec2 = {vec2[0], vec2[1] - mid}; },
      [side, mid](auto &vec2) { vec2 = {mid - vec2[1], side - vec2[0]}; },
      [mid](auto &vec2) { vec2 = {vec2[0] - mid, vec2[1] - mid}; }};

  std::array quads = {std::vector<std::array<double, 2>>{},
                      std::vector<std::array<double, 2>>{},
                      std::vector<std::array<double, 2>>{},
                      std::vector<std::array<double, 2>>{}};

  for (auto &vec2 : vec2s) {
    const bool bit_x = vec2[0] > mid;
    const bool bit_y = vec2[1] > mid;

    const uint8_t quad =
        (static_cast<int>(bit_x) << 1) + static_cast<int>(bit_y);

    maps[quad](vec2);
    quads[quad].push_back(vec2);
  }

  for (auto &quad_vec2s : quads) {
    run_sort_2d(quad_vec2s, mid);
  }

  std::vector<std::array<double, 2>> result;

  const std::array<std::function<void(std::array<double, 2> &)>, 4> invs{
      [](auto &vec2) { vec2 = {vec2[1], vec2[0]}; },
      [mid](auto &vec2) { vec2 = {vec2[0], vec2[1] + mid}; },
      [side, mid](auto &vec2) { vec2 = {side - vec2[1], mid - vec2[0]}; },
      [mid](auto &vec2) { vec2 = {vec2[0] + mid, vec2[1] + mid}; }};

  for (const auto quad : gray2) {
    for (auto &vec2 : quads[quad]) {
      invs[quad](vec2);
    }

    result.insert(result.end(), make_move_iterator(quads[quad].begin()),
                  make_move_iterator(quads[quad].end()));
  }

  vec2s = result;
}

auto is_base_2d(const std::vector<std::array<double, 2>> &vec2s) -> bool {
  if (vec2s.size() < 2) {
    return true;
  }

  const auto &first = vec2s[0];

  return std::all_of(vec2s.begin() + 1, vec2s.end(),
                     [&first](const auto &vec2) { return vec2 == first; });
}

void sort_3d(std::vector<std::array<double, 3>> &vec3s) {
  const auto [min_x_it, max_x_it] = std::ranges::minmax_element(
      vec3s, {}, [](const auto &vec) { return vec[0]; });
  const double min_x = (*min_x_it)[0];
  const double max_x = (*max_x_it)[0];

  const auto [min_y_it, max_y_it] = std::ranges::minmax_element(
      vec3s, {}, [](const auto &vec) { return vec[1]; });
  const double min_y = (*min_y_it)[1];
  const double max_y = (*max_y_it)[1];

  const auto [min_z_it, max_z_it] = std::ranges::minmax_element(
      vec3s, {}, [](const auto &vec) { return vec[2]; });
  const double min_z = (*min_z_it)[2];
  const double max_z = (*max_z_it)[2];

  const double side_x = max_x - min_x;
  const double side_y = max_y - min_y;
  const double side_z = max_z - min_z;
  const double bound = std::max({side_x, side_y, side_z});

  const double scale_x = (side_x == 0 || bound == 0) ? 1 : bound / side_x;
  const double scale_y = (side_y == 0 || bound == 0) ? 1 : bound / side_y;
  const double scale_z = (side_z == 0 || bound == 0) ? 1 : bound / side_z;

  for (auto &vec3 : vec3s) {
    vec3[0] = scale_x * (vec3[0] - min_x);
    vec3[1] = scale_y * (vec3[1] - min_y);
    vec3[2] = scale_z * (vec3[2] - min_z);
  }

  run_sort_3d(vec3s, bound);

  for (auto &vec3 : vec3s) {
    vec3[0] = vec3[0] / scale_x + min_x;
    vec3[1] = vec3[1] / scale_y + min_y;
    vec3[2] = vec3[2] / scale_z + min_z;
  }
};

void run_sort_3d(std::vector<std::array<double, 3>> &vec3s, const double side) {
  if (is_base_3d(vec3s)) {
    return;
  }

  const double mid = side / 2;

  const std::array<std::function<void(std::array<double, 3> &)>, 8> maps{
      [](auto &vec3) { vec3 = {vec3[2], vec3[0], vec3[1]}; },
      [mid](auto &vec3) { vec3 = {vec3[1], vec3[2] - mid, vec3[0]}; },
      [side, mid](auto &vec3) {
        vec3 = {vec3[0], side - vec3[1], mid - vec3[2]};
      },
      [mid](auto &vec3) { vec3 = {vec3[1] - mid, vec3[2] - mid, vec3[0]}; },

      [side, mid](auto &vec3) {
        vec3 = {mid - vec3[2], side - vec3[0], vec3[1]};
      },
      [side, mid](auto &vec3) {
        vec3 = {mid - vec3[1], vec3[2] - mid, side - vec3[0]};
      },
      [side, mid](auto &vec3) {
        vec3 = {vec3[0] - mid, side - vec3[1], mid - vec3[2]};
      },
      [side, mid](auto &vec3) {
        vec3 = {side - vec3[1], vec3[2] - mid, side - vec3[0]};
      }};

  std::array octs = {std::vector<std::array<double, 3>>{},
                     std::vector<std::array<double, 3>>{},
                     std::vector<std::array<double, 3>>{},
                     std::vector<std::array<double, 3>>{},
                     std::vector<std::array<double, 3>>{},
                     std::vector<std::array<double, 3>>{},
                     std::vector<std::array<double, 3>>{},
                     std::vector<std::array<double, 3>>{}};

  for (auto &vec3 : vec3s) {
    const bool bit_x = vec3[0] > mid;
    const bool bit_y = vec3[1] > mid;
    const bool bit_z = vec3[2] > mid;

    const uint8_t oct = (static_cast<int>(bit_x) << 2) +
                        (static_cast<int>(bit_y) << 1) +
                        static_cast<int>(bit_z);

    maps[oct](vec3);
    octs[oct].push_back(vec3);
  }

  for (auto &oct_vec3s : octs) {
    run_sort_3d(oct_vec3s, mid);
  }

  std::vector<std::array<double, 3>> result;

  const std::array<std::function<void(std::array<double, 3> &)>, 8> invs{
      [](std::array<double, 3> &vec3) { vec3 = {vec3[1], vec3[2], vec3[0]}; },
      [mid](std::array<double, 3> &vec3) {
        vec3 = {vec3[2], vec3[0], vec3[1] + mid};
      },
      [side, mid](std::array<double, 3> &vec3) {
        vec3 = {vec3[0], side - vec3[1], mid - vec3[2]};
      },
      [mid](std::array<double, 3> &vec3) {
        vec3 = {vec3[2], vec3[0] + mid, vec3[1] + mid};
      },

      [side, mid](std::array<double, 3> &vec3) {
        vec3 = {side - vec3[1], vec3[2], mid - vec3[0]};
      },
      [side, mid](std::array<double, 3> &vec3) {
        vec3 = {side - vec3[2], mid - vec3[0], vec3[1] + mid};
      },
      [side, mid](std::array<double, 3> &vec3) {
        vec3 = {vec3[0] + mid, side - vec3[1], mid - vec3[2]};
      },
      [side, mid](std::array<double, 3> &vec3) {
        vec3 = {side - vec3[2], side - vec3[0], vec3[1] + mid};
      }};

  for (const auto oct : gray3) {
    for (auto &vec3 : octs[oct]) {
      invs[oct](vec3);
    }

    result.insert(result.end(), make_move_iterator(octs[oct].begin()),
                  make_move_iterator(octs[oct].end()));
  }

  vec3s = result;
}

auto is_base_3d(const std::vector<std::array<double, 3>> &vec3s) -> bool {
  if (vec3s.size() < 2) {
    return true;
  }

  const auto &first = vec3s[0];

  return std::all_of(vec3s.begin() + 1, vec3s.end(),
                     [&first](const auto &vec3) { return vec3 == first; });
}